File: src/oraculus_di_auditor/analysis/signature_chain.py
"""Signature Chain Detector.

Detects unsigned, partially signed, or placeholder-signed compliance documents
in a contract chain. Flags cases where a contract instrument contains signature
gap indicators or unresolved placeholder tokens, which may indicate the document
was executed without valid assent from one or more required parties.
"""

from __future__ import annotations

import re
from typing import Any

from .text_utils import extract_text_content

# Patterns indicating a blank or missing signature block
SIGNATURE_GAP_PATTERNS = [
    r"\bsignature\s+block\s+blank\b",
    r"\bunsigned\b",
    r"\bnot\b.{0,20}\bexecuted\b",
    r"\bplaceholder\b",
    r"_{5,}",  # five or more underscores (blank signature line)
    r"\bDocuSign\b.{0,80}\bpending\b",  # DocuSign envelope not yet completed
    r"\bone\s+party\s+signed\b",
    r"\bagency\s+signature\s+missing\b",
    r"\bvendor\s+signature\s+only\b",
    r"\bcity\s+signature\s+blank\b",
]

# Literal placeholder tokens used by contract-assembly tools
PLACEHOLDER_PATTERN = re.compile(
    r"\\s1\\|\\d1\\|\[SIGNATURE\]",
    re.IGNORECASE,
)

# Contract instrument keywords
CONTRACT_INSTRUMENT_PATTERN = re.compile(
    r"\b(?:MSPA|MSA|PSA|SOW|MOU|agreement|contract|amendment|order\s+form)\b",
    re.IGNORECASE,
)

# Fiscal amount pattern (reused from fiscal layer)
_FISCAL_AMOUNT_PATTERN = re.compile(
    r"\$\s*\d{1,3}(?:,\d{3})*(?:\.\d{2})?|\$\s*\d+(?:\.\d+)?\s*[MBT](?:illion)?",
    re.IGNORECASE,
)

# Compiled signature gap patterns (IGNORECASE applied individually)
_COMPILED_GAP_PATTERNS = [
    re.compile(p, re.IGNORECASE) for p in SIGNATURE_GAP_PATTERNS
]


def _detect_signature_gaps(text: str) -> list[str]:
    """Return each gap-pattern string that matched in *text*."""
    matched = []
    for pattern in _COMPILED_GAP_PATTERNS:
        if pattern.search(text):
            matched.append(pattern.pattern)
    return matched


def _nearest_instrument(text: str, gap_match: re.Match) -> str | None:  # type: ignore[type-arg]
    """Return the contract instrument keyword closest to *gap_match* in *text*."""
    gap_pos = gap_match.start()
    best_word: str | None = None
    best_dist = float("inf")
    for m in CONTRACT_INSTRUMENT_PATTERN.finditer(text):
        dist = abs(m.start() - gap_pos)
        if dist < best_dist:
            best_dist = dist
            best_word = m.group(0).upper()
    return best_word


def detect_signature_anomalies(doc: dict[str, Any]) -> list[dict[str, Any]]:
    """Identify signature chain anomalies in a normalized document.

    Args:
        doc: Normalized document dict

    Returns:
        List of anomaly records; empty if none found.
    """
    anomalies: list[dict[str, Any]] = []

    if not isinstance(doc, dict):
        return anomalies

    text_content = extract_text_content(doc)
    if not text_content:
        return anomalies

    has_instrument = bool(CONTRACT_INSTRUMENT_PATTERN.search(text_content))
    dollar_amounts = _FISCAL_AMOUNT_PATTERN.findall(text_content)

    # Check 1: Signature gap patterns near a contract instrument keyword
    if has_instrument:
        gap_matches: list[re.Match] = []  # type: ignore[type-arg]
        gap_types: list[str] = []
        for pattern in _COMPILED_GAP_PATTERNS:
            m = pattern.search(text_content)
            if m:
                gap_matches.append(m)
                gap_types.append(pattern.pattern)

        if gap_matches:
            # Use the first match to find the nearest instrument keyword
            instrument_type = _nearest_instrument(text_content, gap_matches[0]) or "UNKNOWN"
            severity = "critical" if dollar_amounts else "high"
            anomalies.append(
                {
                    "id": "signature:unsigned-instrument",
                    "issue": (
                        f"Signature gap detected in {instrument_type} instrument"
                    ),
                    "severity": severity,
                    "layer": "signature",
                    "details": {
                        "instrument_type": instrument_type,
                        "signature_gap_type": gap_types[0],
                        "gap_pattern_count": len(gap_matches),
                        "dollar_amount": dollar_amounts[0] if dollar_amounts else None,
                    },
                }
            )

    # Check 2: Placeholder tokens anywhere in the document
    placeholder_match = PLACEHOLDER_PATTERN.search(text_content)
    if placeholder_match:
        anomalies.append(
            {
                "id": "signature:placeholder-tokens",
                "issue": "Unresolved signature placeholder token found in document",
                "severity": "high",
                "layer": "signature",
                "details": {
                    "token": placeholder_match.group(0),
                    "position": placeholder_match.start(),
                },
            }
        )

    return anomalies

File: mobile/lib/analysis/detectors/signatureChain.ts
/**
 * Signature Chain Detector.
 *
 * Detects unsigned, partially signed, or placeholder-signed documents.
 * Direct port of src/oraculus_di_auditor/analysis/signature_chain.py
 */

import { Anomaly, NormalizedDocument } from '../types';
import { extractTextContent } from '../textUtils';

/** Patterns indicating a blank or missing signature block */
export const SIGNATURE_GAP_PATTERNS: string[] = [
  '\\bsignature\\s+block\\s+blank\\b',
  '\\bunsigned\\b',
  '\\bnot\\b.{0,20}\\bexecuted\\b',
  '\\bplaceholder\\b',
  '_{5,}',
  '\\bDocuSign\\b.{0,80}\\bpending\\b',
  '\\bone\\s+party\\s+signed\\b',
  '\\bagency\\s+signature\\s+missing\\b',
  '\\bvendor\\s+signature\\s+only\\b',
  '\\bcity\\s+signature\\s+blank\\b',
];

/** Literal placeholder tokens used by contract-assembly tools */
export const PLACEHOLDER_PATTERN = /\\s1\\|\\d1\\|\[SIGNATURE\]/i;

/** Contract instrument keywords */
export const CONTRACT_INSTRUMENT_PATTERN =
  /\b(?:MSPA|MSA|PSA|SOW|MOU|agreement|contract|amendment|order\s+form)\b/i;

/** Fiscal amount pattern (reused from fiscal layer) */
const FISCAL_AMOUNT_PATTERN =
  /\$\s*\d{1,3}(?:,\d{3})*(?:\.\d{2})?|\$\s*\d+(?:\.\d+)?\s*[MBT](?:illion)?/gi;

/** Compiled signature gap patterns */
const COMPILED_GAP_PATTERNS: RegExp[] = SIGNATURE_GAP_PATTERNS.map(
  (p) => new RegExp(p, 'i')
);

/**
 * Return each gap-pattern string that matched in text.
 * Mirrors Python: _detect_signature_gaps(text)
 */
function detectSignatureGaps(text: string): string[] {
  const matched: string[] = [];
  for (const pattern of COMPILED_GAP_PATTERNS) {
    if (pattern.test(text)) {
      matched.push(pattern.source);
    }
  }
  return matched;
}

/**
 * Return the contract instrument keyword closest to gapPosition in text.
 * Mirrors Python: _nearest_instrument(text, gap_match)
 */
function nearestInstrument(text: string, gapPosition: number): string | null {
  let bestWord: string | null = null;
  let bestDist = Infinity;

  const globalPattern = new RegExp(CONTRACT_INSTRUMENT_PATTERN.source, 'gi');
  let match: RegExpExecArray | null;
  while ((match = globalPattern.exec(text)) !== null) {
    const dist = Math.abs(match.index - gapPosition);
    if (dist < bestDist) {
      bestDist = dist;
      bestWord = match[0].toUpperCase();
    }
  }

  return bestWord;
}

/**
 * Identify signature chain anomalies in a normalized document.
 * Mirrors Python: detect_signature_anomalies(doc)
 */
export function detectSignatureAnomalies(doc: NormalizedDocument): Anomaly[] {
  const anomalies: Anomaly[] = [];

  if (!doc || typeof doc !== 'object') {
    return anomalies;
  }

  const textContent = extractTextContent(doc);
  if (!textContent) {
    return anomalies;
  }

  const hasInstrument = CONTRACT_INSTRUMENT_PATTERN.test(textContent);
  FISCAL_AMOUNT_PATTERN.lastIndex = 0;
  const dollarAmounts = textContent.match(FISCAL_AMOUNT_PATTERN) || [];

  // Check 1: Signature gap patterns near a contract instrument keyword
  if (hasInstrument) {
    const gapMatches: Array<{ index: number; pattern: string }> = [];
    const gapTypes: string[] = [];

    for (const pattern of COMPILED_GAP_PATTERNS) {
      const m = pattern.exec(textContent);
      if (m) {
        gapMatches.push({ index: m.index, pattern: pattern.source });
        gapTypes.push(pattern.source);
      }
      // Reset for stateless patterns
      pattern.lastIndex = 0;
    }

    if (gapMatches.length > 0) {
      const instrumentType =
        nearestInstrument(textContent, gapMatches[0].index) || 'UNKNOWN';
      const severity = dollarAmounts.length > 0 ? 'critical' : 'high';
      anomalies.push({
        id: 'signature:unsigned-instrument',
        issue: `Signature gap detected in ${instrumentType} instrument`,
        severity: severity as 'critical' | 'high',
        layer: 'signature',
        details: {
          instrument_type: instrumentType,
          signature_gap_type: gapTypes[0],
          gap_pattern_count: gapMatches.length,
          dollar_amount: dollarAmounts.length > 0 ? dollarAmounts[0] : null,
        },
      });
    }
  }

  // Check 2: Placeholder tokens anywhere in the document
  const placeholderMatch = PLACEHOLDER_PATTERN.exec(textContent);
  if (placeholderMatch) {
    anomalies.push({
      id: 'signature:placeholder-tokens',
      issue: 'Unresolved signature placeholder token found in document',
      severity: 'high',
      layer: 'signature',
      details: {
        token: placeholderMatch[0],
        position: placeholderMatch.index,
      },
    });
  }

  return anomalies;
}
